pkg/standalone-utils/contracts/BatchRelayer.sol

```diff
@@ -0,0 +1,168 @@
+        return address(uint256(poolId) >> (12 * 8));
+    }
+
+    function joinAndSwap(
+        bytes32 poolId,
+        address payable recipient,
+        IVault.JoinPoolRequest calldata request,
+        IVault.BatchSwapStep[] memory swaps,
+        IAsset[] calldata assets,
+        int256[] calldata limits,
+        uint256 deadline
+    ) external payable returns (int256[] memory swapAmounts) {
+        getVault().joinPool{ value: msg.value }(poolId, msg.sender, address(this), request);
+
+        // Ensure that the BPT gained from the join is all used in the swap
+        require(assets[swaps[0].assetInIndex] == IAsset(_getPoolAddress(poolId)), "Must use BPT as input to swap");
+
+        IERC20 bpt = IERC20(_getPoolAddress(poolId));
+        uint256 bptAmount = bpt.balanceOf(address(this));
+
+        // If necessary, give Vault allowance to take BPT
+        if (bpt.allowance(address(this), address(getVault())) < bptAmount) {
+            bpt.approve(address(getVault()), type(uint256).max);
+        }
+
+        // Ensure that all BPT gained from join is used as input to swap
+        swaps[0].amount = bptAmount;
+
+        // Feed BPT into a GIVEN_IN batch swap.
+        IVault.FundManagement memory funds = IVault.FundManagement({
+            sender: address(this),
+            fromInternalBalance: false,
+            recipient: recipient,
+            toInternalBalance: false
+        });
+
+        swapAmounts = getVault().batchSwap(IVault.SwapKind.GIVEN_IN, swaps, assets, funds, limits, deadline);
+        _sweepETH();
+    }
+
+    function joinAndStake(
+        bytes32 poolId,
+        address payable recipient,
+        IVault.JoinPoolRequest calldata joinPoolRequest
+    ) external payable {
+        getVault().joinPool{ value: msg.value }(poolId, msg.sender, address(this), joinPoolRequest);
+
+        IERC20 bpt = IERC20(_getPoolAddress(poolId));
+        uint256 bptAmount = bpt.balanceOf(address(this));
+
+        // If necessary, give staking contract allowance to take BPT
+        if (bpt.allowance(address(this), address(getStakingContract())) < bptAmount) {
+            bpt.approve(address(getStakingContract()), type(uint256).max);
+        }
+
+        getStakingContract().stake(bpt, bptAmount, recipient);
+        _sweepETH();
+    }
+
+    function swapAndExit(
+        bytes32 poolId,
+        address payable recipient,
+        IVault.ExitPoolRequest memory request,
+        IVault.SwapKind kind,
+        IVault.BatchSwapStep[] calldata swaps,
+        IAsset[] calldata assets,
+        int256[] calldata limits,
+        uint256 deadline
+    ) external payable {
+        // We can't output tokens to the user's internal balance
+        // as they need to have BPT on their address for the exit
+        IVault.FundManagement memory funds = IVault.FundManagement({
+            sender: msg.sender,
+            fromInternalBalance: false,
+            recipient: msg.sender,
```

Comment:
I'd use `address(this)` - both will work, but this seems more direct and less cryptic by not relying on relayer powers for the exit.

Reply:
The reasoning behind this is that we can ensure that should the user provide a swap which results in an output of other tokens in addition to the BPT then this will not be locked on the relayer.